### PR TITLE
[FLINK-15455][network] Enabled tcp connection reuse across multi jobs.

### PR DESCRIPTION
**Pull request by KarmaGYZ**
## What is the purpose of the change

Enabled tcp connection reuse across multi jobs.

## Verifying this change

This change added tests and can be verified as follows:

- NettyPartitionRequestClientTest#testPartitionRequestClientReuse

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by KarmaGYZ** [comment]
cc @xintongsong @wsry

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/AtomicDisposableReferenceCounter.java [hunk 1]
```diff
@@ -94,9 +100,9 @@ public boolean isDisposed() {
         }
     }
 
-    public boolean disposeIfNotUsed() {
+    public boolean disposeIfNotUsed(boolean canDispose) {
         synchronized (lock) {
-            if (referenceCount <= disposeOnReferenceCount) {
+            if (referenceCount <= disposeOnReferenceCount && canDispose) {
                 isDisposed = true;
             }
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/util/AtomicDisposableReferenceCounter.java:107]: I don't know what this class is introduced for in the very beginning, but now it does not make sense to me any more.
- The constructor with custom `disposeOnReferenceCount` is only use in a test case.
- Without custom `disposeOnReferenceCount`, this class basically an `AtomicInteger`, where `isDisposed` can be replaced by `== 0`.
- This class introduces a constraint that the reference count can only reach the "disposed state" once. This IMO is a design flaw. It should be the caller's decision what does the counter reaching zero mean and what action to take at the time.

In summary, I think the code will be significantly simplified if we get rid of this class and replace it with a simple `AtomicInteger`, or a light-weighted wrap with shortcuts like `incAndIsZero`, `decAndIsZero`, `isZero`.

**Review by xintongsong — CHANGES_REQUESTED** [review]
Thanks @KarmaGYZ. I've went through the production codes and have some comments mostly concerning the code complexity and readability. I'd prefer to look into the testing codes later, as they are likely to be changed while addressing these comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientFactory.java [hunk 2]
```diff
@@ -75,6 +78,11 @@
      */
     NettyPartitionRequestClient createPartitionRequestClient(ConnectionID connectionId)
             throws IOException, InterruptedException {
+        // Proactively close idle connections with error.
+        if (connectionReuseEnabled) {
+            closeErrorChannelConnections();
+        }
```
[the remark this answers is not in the document]
Comment (KarmaGYZ) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientFactory.java:84]: > Which component should be responsible for close the connection normally?

[the remark this answers is not in the document]
Comment (KarmaGYZ) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientFactory.java:84]: > Which component should be responsible for close the connection normally?

**Comment by KarmaGYZ** [comment]
Thanks for the review @xintongsong . PR updated.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientFactory.java [hunk 3]
```diff
@@ -103,6 +105,10 @@ NettyPartitionRequestClient createPartitionRequestClient(ConnectionID connection
             } else {
                 try {
                     client = clientFuture.get();
+                    if (client.canBeDisposed()) {
+                        client.closeConnection();
+                        continue;
+                    }
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientFactory.java:111]: Isn't this duplicating the following?
```
            if (client.incrementReferenceCounter()) {
                return client;
            } else {
                destroyPartitionRequestClient(connectionId, client);
            }
```

Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientFactory.java:111]: I'd suggest to keep the latter, as it decouples from the `if-else`/`try-catch` branches.

Moreover, I'd suggest to rename `incrementReferenceCounter` to `validateClientAndIncrementReferenceCounter`.

**Review by xintongsong — CHANGES_REQUESTED** [review]

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyPartitionRequestClient.java [hunk 4]
```diff
@@ -222,22 +227,33 @@ public void close(RemoteInputChannel inputChannel) throws IOException {
...
-            // Close the TCP connection. Send a close request msg to ensure
-            // that outstanding backwards task events are not discarded.
-            tcpChannel
-                    .writeAndFlush(new NettyMessage.CloseRequest())
-                    .addListener(ChannelFutureListener.CLOSE_ON_FAILURE);
-
-            // Make sure to remove the client from the factory
-            clientFactory.destroyPartitionRequestClient(connectionId, this);
+        if (closeReferenceCounter.getAndUpdate(count -> count > 0 ? count - 1 : count) == 1
+                && !canBeReused()) {
```
[the remark this answers is not in the document]
Comment (KarmaGYZ) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyPartitionRequestClient.java:231]: @wsry WDYT?

[the remark this answers is not in the document]
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyPartitionRequestClient.java:231]: 1. Why would a reference counter reaches 0 for more than once when reusing is disabled?
2. Would there be any damage if a connection is closed for more than once?

[the remark this answers is not in the document]
Comment (KarmaGYZ) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyPartitionRequestClient.java:231]: Fine. It seems the close connection operation is idempotent.

**Comment by KarmaGYZ** [comment]
@xintongsong Thanks for the review. PR updated.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/CreditBasedPartitionRequestClientHandlerTest.java [hunk 5]
```diff
@@ -516,7 +516,7 @@ public void testNotifyCreditAvailableAfterReleased() throws Exception {
         final CreditBasedPartitionRequestClientHandler handler =
                 new CreditBasedPartitionRequestClientHandler();
         final EmbeddedChannel channel = new EmbeddedChannel(handler);
-        final PartitionRequestClient client =
+        final NettyPartitionRequestClient client =
```
Comment (xintongsong) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/CreditBasedPartitionRequestClientHandlerTest.java:519]: This change is no longer needed.

**Review by xintongsong — APPROVED** [review]
Thanks for addressing my comments, @KarmaGYZ. LGTM.

**Comment by KarmaGYZ** [comment]
@flinkbot run azure
